File: Erlang/src/Lib/Math/math.erl
-module(math).
-include_lib("eunit/include/eunit.hrl").
-export([fib/1,fact/1,inductive/3,inductive_fact/1,inductive_sum_int/1,inductive_sum_sqr/1]).


fib(0) -> 0;
fib(1) -> 1;
fib(N) -> fib(N-1) + fib(N-2).


fact(0) -> 1;
fact(N) -> N * fact(N - 1).


times(N, M) -> N * M.
double(N) -> times(N, 2).


inductive(Base, _, 0) -> Base;
inductive(Base, Comb, N) -> Comb (N, inductive(Base, Comb, N - 1)).

inductive_fact(N) -> inductive(1, fun (X, Y) -> X * Y end, N).
inductive_sum_int(N) -> inductive(0, fun (X, Y) -> X + Y end, N).
inductive_sum_sqr(N) -> inductive(0, fun (X, Y) -> X * X + Y end, N).

%
% TESTS
%

fib_0_test() -> 0 = fib(0).
fib_1_test() -> 1 = fib(1).
fib_5_test() -> 5 = fib(5).
fib_20_test() -> 6765 = fib(20).
times_0_a_test() -> 0 = times(0, 5).
times_a_0_test() -> 0 = times(5, 0).
times_5_10_test() -> 50 = times(5, 10).
double_5_test() -> 10 = double(5).
inductive_fact_test() -> 120 = inductive_fact(5).
inductive_sum_int_test() -> 210 = inductive_sum_int(20).
inductive_sum_sqr_test() -> 2870 = inductive_sum_sqr(20).

File: Prolog/src/Lib/Math/src/math.pl
:- module(math, [fib/2, fact/2, times/3, double/2, inductive/4, inductive_fact/2, inductive_sum_int/2, inductive_sum_sqr/2]).
:- include('/data/lib/crisp/lib/crisp_includes').


describe(fib/2,
 [
  fib(0,0) ,
  fib(1,1) ,
  fib(20,6765)
 ]).

describe(fact/2,
 [
  fact(0,1) ,
  fact(5,120)
 ]).

describe(times/3,
 [
  times(0,5,0) ,
  times(5,0,0) ,
  times(5,10,50)
 ]).

describe(double/2,
 [
  double(5,10)
 ]).

describe(inductive_fact/2,
 [
  inductive_fact(5, 120)
%  inductive_fact(5, once(fact(5, R)))
 ]).

describe(inductive_sum_int/2,
 [
  inductive_sum_int(20, 210)
 ]).

describe(inductive_sun_sumsqr/2,
 [
  inductive_sum_sqr(20, 2870)
 ]).

fib(0, R) :- R is 0.
fib(1, R) :- R is 1.
fib(N, R) :-
 N1 is N - 1, N2 is N - 2,
 fib(N1, R1), fib(N2, R2),
 R is R1 + R2.
 

fact(0, R) :- R is 1.
fact(N, R) :-
 N1 is N - 1,
 fact(N1, R1),
 R is N * R1.


times(N, M, R) :-
 R is N * M.


double(N, R) :-
 times(N, 2, R).

inductive(Base, Comb, 0, R) :- R is Base.
inductive(Base, Comb, N, R) :-
 N1 is N - 1,
 inductive(Base, Comb, N1, R1),
 call(Comb, N, R1, R2),
 R is R2.

add(X, Y, R) :-
 R is X + Y.

mul(X, Y, R) :-
 R is X * Y.

sumsqr(X, Y, R) :-
 R is X * X + Y.

inductive_fact(N, R) :-
 inductive(1, mul, N, R).

inductive_sum_int(N, R) :-
 inductive(0, add, N, R).

inductive_sum_sqr(N, R) :-
 inductive(0, sumsqr, N, R).
